### PR TITLE
Support for more than one disk #68

### LINKED ISSUE #68: Support for more than one disk
We would like to use this awesome plugin to build new templates on vSphere, which is currently a manual process. However, all our current templates use more than one disk, which is not supported at the moment.

Is support for multiple disks on the roadmap? I am thinking of an option such as [disk_additional_size](https://www.packer.io/docs/builders/vmware-iso.html#disk_additional_size) from the Packer vmware-iso builder.

If there are no plans to implement this, I will try to write the code myself. Would you accept patches to include this feature?

Thanks,
Tim

### PR DESCRIPTION
**Pull request by TimVerhoef**
fixes #68 

The following tasks have been completed:
- add multiple disk support to the iso builder
- translate the requested disk type (e.g. thin) to the required internal options for that choice
- provide the option to set the disk type globally (for all disks) and per disk (which overrides the global option)

The following tasks are still pending:
- add support for naming disks
- add multiple disk support to the clone builder
- update documentation (README.md)
- clean up commit history

The following changes were done for testing and should be reverted before merging:
- change from SATA to IDE controller for the cd-rom
- hash out all build options except for the iso builder on Linux

Also, code for an internal HTTP server was added (based on other plugins that already support it). When this is finished, it will be presented in a seperate pull request.

### CODE REVIEW

**Comment by wcampos** [comment]
Any updates on this feature? :)

**Comment by dkirrane** [comment]
Any plans to merge this PR?

**Comment by NasAmin** [comment]
Hi,

Any ETA on this being merged and released please?

**Comment by lequem** [comment]
Please, merge this PR!

**Comment by thor** [comment]
@lequem there are some things that still remain, including cleaning up the commit history, as well as updating the documentation, according to the pull request. In other words, it's still a *work in progress* :construction: 

Though, whether the original PR is still actually being updated, is... a different question.

**Comment by gowthamakanthan** [comment]
Team, Do this feature is enabled? Please update.

**Comment by dkirrane** [comment]
Wondering is this PR can be closed off?
Also, wondering if this will also address the clone issue https://github.com/jetbrains-infra/packer-builder-vsphere/pull/113 if the VM template has multiple disks?

**Comment by aaronk1** [comment]
@TimVerhoef can you fix conflicts?

**Comment by aaronk1** [comment]
+1 for this functionality

**Comment by pikum** [comment]
+1

**Comment by zyra83** [comment]
+1 for this functionality to.

**Comment by AlekseyUsov** [comment]
+1
Please merge this PR ASAP. We need to attach 3 disks to our VMs and this plugin is the only one that supports vsphere.

**Comment by donwlewis** [comment]
+1
We could really use this feature.  Willing to help out with testing if needed.

**Comment by joezollo** [comment]
Let's get this merged, I am also happy to help. Would be a huge benefit.

**Comment by donwlewis** [comment]
If anyone is interested in a hack/workaround, you can leverage the [govc](https://github.com/vmware/govmomi/blob/master/govc/USAGE.md) binary, and then run something like this to add your disk at the provisioning stage:

```json
{
      "type": "shell-local",
      "command": "govc vm.disk.create -vm {{user `template_name`}} -name {{user `template_name`}}/{{user `template_name`}}_1 -size 10G -ds {{user `datastore`}}"
}
```

**Comment by AlekseyUsov** [comment]
Anyone? @TimVerhoef? Many people have been waiting for more than a year to see this feature, myself included. How can we help to speed this up? This is taking really long.

**Comment by AlekseyUsov** [comment]
Thank you!

File: clone/step_clone.go [hunk 1]
```diff
@@ -10,11 +10,12 @@ import (
 )
 
 type CloneConfig struct {
-	Template    string `mapstructure:"template"`
-	DiskSize    int64  `mapstructure:"disk_size"`
-	LinkedClone bool   `mapstructure:"linked_clone"`
-	Network     string `mapstructure:"network"`
-	Notes       string `mapstructure:"notes"`
+	DiskSize      int64    `mapstructure:"disk_size"`
```
Comment (VladRassokhin) [inline clone/step_clone.go]: It's hard to track difference if parameters are rearranged

File: driver/vm.go [hunk 2]
```diff
@@ -490,26 +498,52 @@ func addDisk(_ *Driver, devices object.VirtualDeviceList, config *CreateConfig)
...
+	for _, dc := range config.Storage {
+		if dc.DiskName != "" {
+			dc.DiskName += ".vmdk"
+		}
+
+		if dc.DiskType == "" && config.GlobalDiskType != "" {
+			dc.DiskType = config.GlobalDiskType
+		}
+
+		if dc.DiskType == "thin" {
```
Comment (VladRassokhin) [inline driver/vm.go]: There should be input (`DiskType`) validation in some other place. 
Probably `DiskEagerlyScrub` and `DiskThinProvisioned` cloud be package-private (lowerCase)

**Review by VladRassokhin — CHANGES_REQUESTED** [review]
Most major blocking point: missing backward compatibility. I'd expect old configs to be still working once this merged.
Also some tests could be useful.

Please also rebase on top of master and reduce number of commits to some reasonable number

File: clone/step_clone.go (hunk 1 above)
Comment (TimVerhoef) [reply clone/step_clone.go]: True. I will undo the rearrangements for clarity.

File: driver/vm.go (hunk 2 above)
Comment (TimVerhoef) [reply driver/vm.go]: Thanks, I will look into this.

**Comment by TimVerhoef** [comment]
> Thank you! Can someone merge it, please?
This issue is not yet ready to merge. There is still a lot of work to do. The plan is to have it ready this week though.

**Comment by AlekseyUsov** [comment]
> > Thank you! Can someone merge it, please?
> 
> This issue is not yet ready to merge. There is still a lot of work to do. The plan is to have it ready this week though.

Thank you for keeping everyone updated.

**Comment by TimVerhoef** [comment]
Due to the

**Comment by TimVerhoef** [comment]
Apologies for the accidental close of the PR. I have cleaned up all commit history, because a lot has changed in the past year and the past commits are not relevant anymore. For reference I have copied the old branch into branch '20190826'.

File: driver/vm.go [hunk 3]
```diff
@@ -546,27 +546,27 @@ func addDisks(_ *Driver, devices object.VirtualDeviceList, config *CreateConfig)
...
 		} else if dc.DiskType == "thick_lazy" {
-			dc.DiskEagerlyScrub    = false
-			dc.DiskThinProvisioned = false
+			dc.diskEagerlyScrub    = false
+			dc.diskThinProvisioned = false
 		} else {
 			// default disk type: Thick provisioned lazy zeroed
-			dc.DiskEagerlyScrub    = false
-			dc.DiskThinProvisioned = false
+			dc.diskEagerlyScrub    = false
```
Comment (VladRassokhin) [inline driver/vm.go]: Let's merge that branch with previous `thick_lazy`. Also I'd recommend to move whole if-else to `CreateConfig#Prepare` which is called from 'prepare' state of step

Comment (TimVerhoef) [reply driver/vm.go]: The reason I wrote it like this, is to have a default value for the disk type and also explicitly naming all the possible options. Also, if the default is ever changed, the values for thick_lazy are already there. But I agree it is not strictly necessary to keep it like this.

I don't know if this conditional can be moved elsewhere, I would have to look into that. The reason it is here is to be able to specify a disk type per individual disk.

**Comment by TimVerhoef** [comment]
The plugin has been re-implemented on top of the lastest version. For now multiple disk- and network support is only implemented for the iso builder. I will try and have a look at the clone builder to see what needs to be done there, but I assume most people wanting this feature use the iso builder.

Also, I did not succeed in adding support for naming disks. If anyone has an idea how to do that I will have a look.

The commit history has been cleaned up and I have updated the README.

File: driver/vm.go (hunk 3 above)
Comment (TimVerhoef) [reply driver/vm.go]: The disk type logic has been moved to function CreateConfig

**Comment by TimVerhoef** [comment]
@VladRassokhin I think this pull request is ready to be merged. If there are any additional requirements please let me know.
Support for multiple disks for the clone builder is not implemented yet, I will try to work on it this week (time permitting). If I can make it work I will open a new pull request.

**Comment by cathode911** [comment]
Colleagues, thank you for great work you've done!
How can we help to speed this up? Lack of multiple disk support is really a blocker for my project

**Comment by TimVerhoef** [comment]
> Colleagues, thank you for great work you've done!
> How can we help to speed this up? Lack of multiple disk support is really a blocker for my project

Do you need multiple disk support for the iso builder or the clone builder? The first one is done and waiting to be merged, the second one is still being worked on (time permitting).

**Comment by cathode911** [comment]
> > Colleagues, thank you for great work you've done!
> > How can we help to speed this up? Lack of multiple disk support is really a blocker for my project
> 
> Do you need multiple disk support for the iso builder or the clone builder? The first one is done and waiting to be merged, the second one is still being worked on (time permitting).

For us multiple disk support for the iso builder is needed.

**Comment by cathode911** [comment]
Any progress so far?

**Comment by TimVerhoef** [comment]
> Any progress so far?
The pull request is ready to be merged to master. I don't know why it hasn't happened yet.

**Comment by cathode911** [comment]
> > Any progress so far?
> 
> The pull request is ready to be merged to master. I don't know why it hasn't happened yet.

@VladRassokhin could you please comment on that?

**Comment by VladRassokhin** [comment]
@cathode911 That's no a project I'm working on, just some side-story, so I'll take a look once I've time. Meanwhile you could built it and try.

**Comment by zpandey** [comment]
> @cathode911 That's no a project I'm working on, just some side-story, so I'll take a look once I've time. Meanwhile you could built it and try.

Any Progress on this?

**Comment by bhundven** [comment]
This builder has been merged upstream in packer in 1.5.2:
https://github.com/hashicorp/packer/pull/8480

This repo is dead. Open issues or merge requests upstream in packer:
https://github.com/hashicorp/packer

**Comment by bhundven** [comment]
https://github.com/hashicorp/packer/issues/8749

**Comment by VladRassokhin** [comment]
This repo is deprecated. AFAIK this change is already merged in [official Packer repository](https://github.com/hashicorp/packer).

**Pull request closed without merging**